### PR TITLE
implementation of xdg-foreign protocol

### LINKED ISSUE #1702: Add xdg-foreign support
This is used by flatpak applications, and potentially others.

### PR DESCRIPTION
**Pull request by ammen99**
Supersedes #1841 (for now only rebased) and thus fixes #1702 

TODO:

- [x] Make xdg-foreign v1 and v2 share resources (as discussed in https://github.com/swaywm/wlroots/pull/1841#issuecomment-721125163)

### CODE REVIEW

**Comment by ammen99** [comment]
I added a new `wlr_xdg_foreign_registry`. The compositors now can choose whether to let v1 and v2 share handles or not.

A basic implementation in a compositor is here: https://github.com/WayfireWM/wayfire/pull/892

File: include/util/uuid.h
Comment (emersion) [inline include/util/uuid.h:4]: For consistency with the rest of wlroots, we could change this to return a `bool`.

File: util/uuid.c
Comment (emersion) [inline util/uuid.c:1]: Nit: local includes should come after external includes.

File: util/meson.build
Comment (emersion) [inline util/meson.build:13]: We try to avoid `#ifdef`s and use `#if` instead so that forgetting to define the macro triggers an error.

File: types/meson.build
Comment (emersion) [inline types/meson.build:70]: Can we move this one line to the main `meson.build`? It's too easy to incorrectly include another Meson file that depends on this being set before executing `subdir('types')`.

File: meson.build
Comment (emersion) [inline meson.build:112]: I think we should have a Meson feature option for xdg-foreign, just like we have for all other features. Package maintainers don't want auto-detection of features, they want hard failures when a dep is missing.

File: include/wlr/types/wlr_xdg_foreign_registry.h
Comment (emersion) [inline include/wlr/types/wlr_xdg_foreign_registry.h:36]: We don't have a destroy listener for this surface. Do we rely on the caller to properly destroy the `wlr_xdg_foreign_exported` when the surface is destoyed?

If so, we should properly document that.

File: include/wlr/types/wlr_xdg_foreign_registry.h
Comment (emersion) [inline include/wlr/types/wlr_xdg_foreign_registry.h:58]: Maybe rename to `wlr_xdg_foreign_exported_init` for consistency with the rest of wlroots?

File: include/wlr/types/wlr_xdg_foreign_registry.h
Comment (emersion) [inline include/wlr/types/wlr_xdg_foreign_registry.h:56]: We don't use `@return` comments.

File: include/wlr/types/wlr_xdg_foreign_registry.h
Comment (emersion) [inline include/wlr/types/wlr_xdg_foreign_registry.h:72]: Maybe rename to `wlr_xdg_foreign_exported_fini` for consistency with the rest of wlroots?

File: include/wlr/types/wlr_xdg_foreign_v2.h
Comment (emersion) [inline include/wlr/types/wlr_xdg_foreign_v2.h:18]: We don't have a `resources` list for globals anymore.

File: include/wlr/types/wlr_xdg_foreign_v2.h
Comment (emersion) [inline include/wlr/types/wlr_xdg_foreign_v2.h:19]: Do we really need per-client lists? Can't we just have directly a list of `wlr_xdg_exported_v2`/`wlr_xdg_imported_v2`?

File: include/wlr/types/wlr_xdg_foreign_v2.h
Comment (emersion) [inline include/wlr/types/wlr_xdg_foreign_v2.h:83]: We don't expose destructors for globals like this one anymore.

File: types/wlr_xdg_foreign_registry.c
Comment (emersion) [inline types/wlr_xdg_foreign_registry.c:8]: The doc-comment is already in the header file. Would prefer not to duplicate it.

File: types/wlr_xdg_foreign_registry.c
Comment (emersion) [inline types/wlr_xdg_foreign_registry.c:27]: Nit: unnecessary newline.

File: types/wlr_xdg_foreign_registry.c
Comment (emersion) [inline types/wlr_xdg_foreign_registry.c:45]: Nit: unnecessary newline.

File: types/wlr_xdg_foreign_v2.c
Comment (emersion) [inline types/wlr_xdg_foreign_v2.c:31]: Hm, why do we need to check whether the surface is a toplevel and send a protocol error if not? The protocol doesn't define any errors like this.

File: types/wlr_xdg_foreign_v2.c
Comment (emersion) [inline types/wlr_xdg_foreign_v2.c:402]: The resources should already be destroyed by the time the `wl_display` is.

**Review by emersion — CHANGES_REQUESTED** [review]
Thanks for continuing the effort on this feature! Here's a first round of comments.

Apart from these comments, I'd also like the commits to be re-organized and split. For instance the xdg-shell change can be in a standalone commit, so can be the introduction of `util/uuid`. You can use [`Co-authored-by` tags](https://docs.github.com/en/free-pro-team@latest/github/committing-changes-to-your-project/creating-a-commit-with-multiple-authors) in the commit message to indicate a secondary commit author.

File: include/wlr/types/wlr_xdg_foreign_registry.h
Comment (ammen99) [reply include/wlr/types/wlr_xdg_foreign_registry.h:58]: I don't like `wlr_xdg_foreign_exported_init` because the registry initializes the given exported surface, it is not like the exported surface initializes itself (if it were C++ I'd put this method inside the registry class). If you however really don't like the name I can rename it, no problem.

Comment (emersion) [reply include/wlr/types/wlr_xdg_foreign_registry.h:58]: I was thinking of changing the order of arguments too:

```c
bool wlr_xdg_foreign_exported_init(struct wlr_xdg_foreign_exported *surface, struct wlr_xdg_foreign_registry *registry);
```

which would mean "initialize the exported foreign given this registry".

I'm also fine with keeping the current order, but I'd really like to use the `init`/`finish` wording we use elsewhere.

File: include/wlr/types/wlr_xdg_foreign_registry.h
Comment (emersion) [inline include/wlr/types/wlr_xdg_foreign_registry.h:72]: Also, we shouldn't need to provide the registry here, `wlr_xdg_foreign_exported` already has a registry field.

File: include/wlr/types/wlr_xdg_foreign_v2.h
Comment (ammen99) [reply include/wlr/types/wlr_xdg_foreign_v2.h:19]: yeah the list is actually of wlr_xdg_exported_v2 not of clients, I think `objects` would be a better name, or what do you think?

**Comment by ammen99** [comment]
Thanks for the review, I'm working on these changes. I will be just pushing new commits until we're ready for merging, and then I'll reorganize the commits as you suggested.

File: include/wlr/types/wlr_xdg_foreign_registry.h
Comment (ammen99) [reply include/wlr/types/wlr_xdg_foreign_registry.h:36]: yeah, I explained in in the docstring for init()

File: types/wlr_xdg_foreign_v2.c
Comment (ammen99) [reply types/wlr_xdg_foreign_v2.c:31]: Should we instead silently fail?

File: meson.build
Comment (ammen99) [reply meson.build]: I have added this, let me know what you think of the approach.

**Comment by ammen99** [comment]
I think I have addressed all of your comments, except the one about erroring when the handle is not a toplevel. The protocol says `Only xdg_surface surfaces may be exported.` which I think is enough grounds to disconnect a client which exports something else.

File: include/util/uuid.h
Comment (emersion) [inline include/util/uuid.h:6]: Maybe we should have a constant for this?

```c
#ifndef UUID_STR_LEN
#define UUID_STR_LEN 37
#endif
```

File: meson.build
Comment (emersion) [inline meson.build:118]: Nit: `set10` takes a `bool`, so should be `true` instead of 1.

File: meson.build
Comment (emersion) [inline meson.build:120]: Nit: missing space in error message.

File: include/util/uuid.h
Comment (emersion) [reply include/util/uuid.h:6]: Hm, I guess it wouldn't help too much, since this is a private header.

File: util/meson.build
Comment (emersion) [inline util/meson.build:20]: Nit: this can just be single-line

File: util/uuid.c
Comment (emersion) [inline util/uuid.c:27]: `strcpy` is insecure. Use `memcpy` instead. Maybe `assert(strlen(str) + 1 == 37)` would be nice too.

File: include/wlr/types/wlr_xdg_foreign_v1.h
Comment (emersion) [inline include/wlr/types/wlr_xdg_foreign_v1.h:20]: We have `exporter_global` and `exporter.global`. Do we need both?

File: types/wlr_xdg_foreign_v1.c
Comment (emersion) [inline types/wlr_xdg_foreign_v1.c:396]: Use NULL as data argument to allow future extensibility.

File: types/wlr_xdg_foreign_v1.c
Comment (emersion) [inline types/wlr_xdg_foreign_v1.c:222]: This is already done by `wlr_xdg_foreign_exported_init`

File: types/wlr_xdg_foreign_v1.c
Comment (emersion) [inline types/wlr_xdg_foreign_v1.c:244]: Nothing in the protocol says that destroying the exporter also destroys the child objects.

File: types/wlr_xdg_foreign_v1.c
Comment (emersion) [inline types/wlr_xdg_foreign_v1.c:357]: Nothing in the protocol says that destroying the importer also destroys the child objects.

File: include/wlr/types/wlr_xdg_foreign_v1.h
Comment (emersion) [inline include/wlr/types/wlr_xdg_foreign_v1.h:36]: I still don't think we need this intermediate struct. We can just store the imported/exported objects in the `wlr_xdg_foreign_v1`.

File: include/wlr/types/wlr_xdg_foreign_v1.h
Comment (ammen99) [reply include/wlr/types/wlr_xdg_foreign_v1.h:20]: Indeed, this is a mistake, will remove.

File: types/wlr_xdg_foreign_v1.c
Comment (ammen99) [reply types/wlr_xdg_foreign_v1.c:222]: It isn't done but probably should be :)

File: include/wlr/types/wlr_xdg_foreign_v1.h
Comment (ammen99) [reply include/wlr/types/wlr_xdg_foreign_v1.h]: You are right! I misunderstood your previous comment. This change removed a lot of unnecessary code.

**Comment by ammen99** [comment]
Alright, made it so that we don't store anything about the intermediate exporter/importer resources. They directly have the `foreign` as their user data, which is enough.

File: types/wlr_xdg_foreign_registry.c
Comment (emersion) [inline types/wlr_xdg_foreign_registry.c:40]: Maybe we should trigger the `destroy` event here?

File: types/wlr_xdg_foreign_v2.c
Comment (emersion) [reply types/wlr_xdg_foreign_v2.c:31]: No, reading the protocol again, sending a protocol error is what we should do, but we're just missing the error values.

https://gitlab.freedesktop.org/wayland/wayland-protocols/-/merge_requests/57

File: types/wlr_xdg_foreign_v2.c
Comment (emersion) [inline types/wlr_xdg_foreign_v2.c:304]: Missing a NULL check here

File: types/wlr_xdg_foreign_v2.c
Comment (emersion) [inline types/wlr_xdg_foreign_v2.c:315]: I'd prefer to make imported objects without an exported inert, by making the user data pointer NULL and free'ing the `struct wlr_xdg_imported_v2`. This is the standard thing that wlroots does when objects are still alive but can't be used anymore.

File: types/wlr_xdg_foreign_v2.c
Comment (emersion) [inline types/wlr_xdg_foreign_v2.c:212]: Missing a NULL check here.

File: types/wlr_xdg_foreign_v2.c
Comment (emersion) [inline types/wlr_xdg_foreign_v2.c:226]: We checked `verify_is_toplevel` above, so this should be unnecessary?

File: types/wlr_xdg_foreign_v2.c
Comment (emersion) [inline types/wlr_xdg_foreign_v2.c:183]: This destroys the resource behind the client's back. We can't do this, otherwise the client will die when sending an `xdg_exported.destroy` request.

Why do we need to handle `unmap` in the first place? The protocol doesn't seem to mention anything about it.

File: types/wlr_xdg_foreign_v2.c
Comment (emersion) [inline types/wlr_xdg_foreign_v2.c:86]: Hm, `verify_is_toplevel` should take care of this too I think?

File: types/wlr_xdg_foreign_v2.c
Comment (emersion) [inline types/wlr_xdg_foreign_v2.c:107]: This should always be true as well I think.

File: types/wlr_xdg_foreign_v2.c
Comment (emersion) [inline types/wlr_xdg_foreign_v2.c:118]: `child` is leaked in the else branch, because `wlr_xdg_toplevel_set_parent` won't free it in `xdg_imported_handle_resource_destroy`.

File: types/wlr_xdg_foreign_v2.c
Comment (ammen99) [reply types/wlr_xdg_foreign_v2.c:86]: Yes, it was a leftover from the time when there was both xdg6 and xdg-stable

File: types/wlr_xdg_foreign_v2.c
Comment (ammen99) [reply types/wlr_xdg_foreign_v2.c:118]: if we remove the if() there is no else, right?

File: types/wlr_xdg_foreign_v2.c
Comment (ammen99) [reply types/wlr_xdg_foreign_v2.c:183]: A good question. I will remove that as well.

Comment (ammen99) [reply types/wlr_xdg_foreign_v2.c:183]: Hmm, I suppose we want to listen not for unmap, but for the case where the toplevel is destroyed.

**Comment by ammen99** [comment]
I think I have addressed your new review comments. Inert state is used in two different places:

- For imported, if the exported could not be found
- For exported, if the xdg-surface is destroyed (not unmapped, as it was previously).

**Review by emersion — COMMENTED** [review]
This looks good to me, thanks. Can you split everything into logical commits? See https://github.com/swaywm/wlroots/pull/2487#pullrequestreview-546948373

**Comment by ammen99** [comment]
> This looks good to me, thanks. Can you split everything into logical commits? See [#2487 (review)](https://github.com/swaywm/wlroots/pull/2487#pullrequestreview-546948373)

Thanks for the review! I have split everything into five commits.

**Review by emersion — APPROVED** [review]
Nice work, thanks for you contribution!

**Pull request merged**
